src/AI/WallyAI.java

```diff
@@ -981,212 +1917,533 @@ private Map<XYCoord, UnitModel> queueUnitProductionActions(GameMap gameMap)
     {
-      // We don't currently have any huge cost-shift COs, so this isn't a big deal at present.
-      ent.setValue(ent.getValue() * ent.getKey().costBase);
+      ModelForCO tmco = ent.getKey();
+      var fundsEntry = new AbstractMap.SimpleEntry<ModelForCO, Integer>(tmco, ent.getValue() * tmco.co.getCost(tmco.um) / 10);
+      enemyModels.add(fundsEntry);
     }
 
-    Queue<Entry<UnitModel, Double>> enemyModels = 
-        new PriorityQueue<Entry<UnitModel, Double>>(myArmy.cos[0].unitModels.size(), new UnitModelFundsComparator());
-    enemyModels.addAll(enemyUnitCounts.entrySet());
-
     // Try to purchase units that will counter the most-represented enemies.
     while (!enemyModels.isEmpty() && !CPI.availableUnitModels.isEmpty())
     {
-      // Find the first (most funds-invested) enemy UnitModel, and remove it. Even if we can't find an adequate counter,
-      // there is not reason to consider it again on the next iteration.
-      UnitModel enemyToCounter = enemyModels.poll().getKey();
-      double enemyNumber = enemyUnitCounts.get(enemyToCounter);
-      log(String.format("Need a counter for %sx%s", enemyToCounter, enemyNumber / enemyToCounter.costBase / UnitModel.MAXIMUM_HEALTH));
-      log(String.format("Remaining budget: %s", budget));
+      Entry<ModelForCO, Integer> enemyEntry = enemyModels.poll();
+      ModelForCO enemyToCounter = enemyEntry.getKey();
+      int enemyHP = enemyUnitHP.get(enemyToCounter);
+      if( enemyHP < 0 ) // Just try to counter tanks if the enemy doesn't have anything "interesting"
+      {
+        enemyToCounter = new ModelForCO(enemyToCounter.co, tankModel);
+        enemyHP        = 42;
+      }
+      log(String.format("Need a counter for %s HP of %s (have %s)", enemyHP, enemyToCounter, budget));
 
       // Get our possible options for countermeasures.
-      ArrayList<UnitModel> availableUnitModels = new ArrayList<>();
-      for( ModelForCO coModel : CPI.availableUnitModels )
-        availableUnitModels.add(coModel.um);
-      while (!availableUnitModels.isEmpty())
+      int costBuffer = 0;
+      ArrayList<ModelForCO> availableUnitModels = new ArrayList<>();
+      for( ModelForCO currentCounter : CPI.availableUnitModels )
       {
-        // Sort my available models by their power against this enemy type.
-        Collections.sort(availableUnitModels, new UnitPowerComparator(enemyToCounter, this));
-
-        // Grab the best counter.
-        UnitModel idealCounter = availableUnitModels.get(0);
-        availableUnitModels.remove(idealCounter); // Make sure we don't try to build two rounds of the same thing in one turn.
         // I only want combat units, since I don't understand transports
-        if( !idealCounter.weapons.isEmpty() )
+        if( currentCounter.um.weapons.isEmpty() )
+          continue;
+        // Filter stuff I can't afford; this breaks if we have localized discounts, but those don't exist (yet)
```

Comment:
We _do_ have localized price increases, if that makes a difference.

Reply:
Though doesn't line 2022 get the location-specific cost anyway?

Reply:
It doesn't, in this case.
If I can't buy a tank at the normal price, it doesn't matter that I can also buy it at 8k.

I also have some thoughts about how I should (re)design his buying code, but distractions are very cruel.
You can imagine why I hadn't put this up for review for so long - there are always more changes I can/should make.

Reply:
> Though doesn't line 2022 get the location-specific cost anyway?

Technically yes, but that's a single location-specific cost (from an effectively arbitrary production location, since another unit type may be chosen to build there first)